File: demo/csep590b_c04/p2_safe_mul/safe_mul_fixed.cpp
/*
DEMO: Problem 2 (safe_mul) — FIXED reference.

Multiply in a wider type (int64_t), check whether the result falls
in the signed-i32 range, return 0 on overflow.

Expected verdict: SAT (VERIFIED) against `safe_mul_spec`.
*/

#include <climits>
#include <cstdint>

int32_t safe_mul(int32_t a, int32_t b) {
    int64_t c = (int64_t)a * (int64_t)b;
    if (c > (int64_t)INT32_MAX || c < (int64_t)INT32_MIN) {
        return 0;
    }
    return (int32_t)c;
}

File: demo/csep590b_c04/p2_safe_mul/safe_mul_fixed.rs
// DEMO: Problem 2 (safe_mul) — FIXED reference (Rust mirror of
// safe_mul_fixed.cpp).
//
// Multiply in i64, check the i32 range, return 0 on overflow.
//
// Expected verdict: VERIFIED against `safe_mul_spec`.

fn safe_mul(a: i32, b: i32) -> i32 {
    let c: i64 = (a as i64) * (b as i64);
    if c > i32::MAX as i64 || c < i32::MIN as i64 {
        0
    } else {
        c as i32
    }
}
